File: apps/youtube/extensions/quality/src/main/java/app/reseam/youtube/quality/RememberVideoQuality.java
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.youtube.quality;

import app.reseam.runtime.settings.ReseamSettings;
import app.reseam.youtube.core.Logger;
import app.reseam.youtube.core.Settings;
import app.reseam.youtube.video.VideoInformation;

/** Applies and stores the selected quality without depending on YouTube's preference UI. */
public final class RememberVideoQuality {
    private RememberVideoQuality() {}

    public static boolean shouldRememberVideoQuality() {
        return Settings.getBoolean("remember_video_quality", false);
    }

    public static int getDefaultQualityResolution() {
        try {
            return Integer.parseInt(Settings.getString("video_quality_default", "-2"));
        } catch (NumberFormatException exception) {
            return VideoInformation.AUTOMATIC_VIDEO_QUALITY_VALUE;
        }
    }

    public static void newVideoStarted(VideoInformation.PlaybackController controller) {
        int resolution = getDefaultQualityResolution();
        VideoInformation.setDesiredVideoResolution(resolution);
        Logger.debug(() -> "Default video quality applied: " + resolution);
    }

    public static void userChangedQuality(int resolution) {
        if (!shouldRememberVideoQuality()) return;
        ReseamSettings.setString("you_tube_settings.video_quality_default", Integer.toString(resolution));
        Logger.debug(() -> "Remembered video quality: " + resolution);
    }

    public static void userChangedShortsQuality(int index) {
        VideoInformation.VideoQualityInterface[] qualities = VideoInformation.getCurrentQualities();
        if (qualities == null || index < 0 || index >= qualities.length || qualities[index] == null) return;
        userChangedQuality(qualities[index].patch_getResolution());
    }
}

File: apps/youtube/patch/src/main/kotlin/app/reseam/patches/youtube/video/quality/RememberVideoQuality.kt
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.patches.youtube.video.quality

import app.reseam.patch.ExtClass
import app.reseam.patch.Type
import app.reseam.patch.before
import app.reseam.patch.dex.AccessFlags
import app.reseam.patch.dex.parameterTypes
import app.reseam.patch.klass
import app.reseam.patch.method
import app.reseam.patch.patch
import app.reseam.patch.point
import app.reseam.patch.settings.section
import app.reseam.patches.youtube.core.YOUTUBE
import app.reseam.patches.youtube.core.YouTubeSettings
import app.reseam.patches.youtube.core.YouTubeSettingsPages
import app.reseam.patches.youtube.core.youTubeSettings
import app.reseam.patches.youtube.internal.onCreateHook
import app.reseam.patches.youtube.internal.setPlaybackSpeedMethod
import app.reseam.patches.youtube.internal.videoInformationHook
import app.reseam.patches.youtube.internal.videoQualityClass

object RememberVideoQuality : ExtClass("app.reseam.youtube.quality.RememberVideoQuality") {
    val newVideoStarted = static("newVideoStarted", "app.reseam.youtube.video.VideoInformation\$PlaybackController")
    val userChangedQuality = static("userChangedQuality", Type.Int)
    val userChangedShortsQuality = static("userChangedShortsQuality", Type.Int)
}

private val qualityMenuParent = method("video quality menu parent") {
    strings("VIDEO_QUALITIES_MENU_BOTTOM_SHEET_FRAGMENT")
    returns(Type.Void)
}

private val qualityMenuClass = app.reseam.patch.classTarget("video quality menu class") {
    bytecode.findClass(qualityMenuParent.owner) ?: error("The video quality menu class is missing")
}

private val qualityItemClick = method("video quality item click") {
    inClass(qualityMenuClass)
    name("onItemClick")
    flags(AccessFlags.PUBLIC or AccessFlags.FINAL)
    returns(Type.Void)
    paramCount(4)
    param(2, Type.Int)
    param(3, Type.Long)
    custom { parameterTypes[0].startsWith("L") && parameterTypes[1].startsWith("L") }
}

private val qualitySetter = method("player quality setter") {
    inClass(klass(setPlaybackSpeedMethod.owner))
    params(videoQualityClass.descriptor)
    returns(Type.Void)
}

private val qualityChangedMethod = method("video quality command") {
    calls(qualitySetter)
    paramCount(1)
    custom { returnType.startsWith("L") }
}

private val qualityChangedValue = qualityChangedMethod.point("selected video quality") {
    invokeDirect { owner(videoQualityClass.descriptor); name("<init>") }
}.captureArgumentAs("quality", 1, Type.Int)

val rememberVideoQuality = patch("Remember video quality") {
    description("Remembers the selected video quality and applies the configured default.")
    compatibleWith(YOUTUBE)
    dependsOn(youTubeSettings, videoInformationHook)
    settings(youTubeSettings, section(YouTubeSettingsPages.Video, "Video quality", YouTubeSettings.videoQualityDefault, YouTubeSettings.rememberVideoQuality))

    execute {
        onCreateHook(RememberVideoQuality.newVideoStarted)
        qualityChangedValue.before { call(RememberVideoQuality.userChangedQuality, capture("quality")) }
        qualityItemClick.before { call(RememberVideoQuality.userChangedShortsQuality, param(2)) }
    }
}
